### PR TITLE
+PDF prev

### PR DESCRIPTION
**Pull request by voroneril**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "ejs": "^3.1.6",
     "express": "^4.16.4",
     "file-loader": "^3.0.1",
+    "gm": "^1.23.1",
```
Comment (RealDolos) [inline package.json:24]: Make this an optional dependency please. I really do not like gm/imagemagick much. And it's a huge dependency.
https://docs.npmjs.com/cli/v7/configuring-npm/package-json#optionaldependencies

File: lib/meta.js [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
 const fs = require("fs");
+const fsPromises = fs.promises;
 const {spawn} = require("child_process");
 const path = require("path");
 const XRegExp = require("xregexp");
 const {PromisePool} = require("./util");
 const sharp = require("sharp");
 const CONFIG = require("./config");
+const gm = require('gm')
```
Comment (RealDolos) [inline lib/meta.js:11]: if you make this optional you have to guard with try-catch here

File: lib/meta.js [hunk 3]
```diff
@@ -195,7 +249,7 @@ async function generateAssetsAudio(storage) {
     exiftool = [
       "firejail",
       "--quiet", `--profile=${PROFILE}`, `--private=${p.dir}`,
-      EXIFTOOL, "-b", "-Picture", `./${p.base}`
+      EXIFTOOL, "-b", "-Picture", `${p.dir}/${p.base}`
```
Comment (RealDolos) [inline lib/meta.js:252]: nope, this was correct, except your jailing is not working. Turn off jail in your config, and do not change this

File: lib/meta.js [hunk 4]
```diff
@@ -107,7 +160,8 @@ async function generateAssetsImage(storage) {
   }
   const assets = [];
   const known = new Set();
-  const base = await sharp(storage.full, {limitInputPixels: PIXEL_LIMIT}).
+  const base = await sharp(storage.full).
```
Comment (RealDolos) [inline lib/meta.js:163]: this breaks on later sharps that this repo uses. Do not change it like this

File: lib/meta.js [hunk 5]
```diff
@@ -227,6 +281,10 @@ async function generateAssetsAudio(storage) {
 
 async function generateAssets(storage) {
   const {mime} = storage;
+  console.log('storage.meta.type', storage.meta.type);
```
Comment (RealDolos) [inline lib/meta.js:284]: make this a console.debug

File: lib/meta.js [hunk 6]
```diff
@@ -227,6 +281,10 @@ async function generateAssetsAudio(storage) {
...
+  if (storage.meta.type == "PDF" ||storage.meta.type == "pdf") {
```
Comment (RealDolos) [inline lib/meta.js:285]: Optional check ;)

File: lib/meta.js [hunk 7]
```diff
@@ -99,6 +101,57 @@ async function runcmd(cmdargs, encoding) {
...
+  const assets = [];
+  const known = new Set();
+  const pdfBuffer = await fsPromises.readFile(storage.full);
+  const prev = await gm(pdfBuffer)
+    .selectFrame(0) 
+    .size()
+    .toBuffer('jpg',async function (err, prev) {
+      
+      const base = await sharp(prev).
+        limitInputPixels(PIXEL_LIMIT).
```
Comment (RealDolos) [inline lib/meta.js:114]: Same issue here that later versions of sharp does not support it like this...

File: lib/meta.js [hunk 8]
```diff
@@ -250,7 +308,7 @@ async function detectMime(file) {
       filetool = [
         "firejail",
         "--quiet", `--profile=${PROFILE}`, `--private=${p.dir}`,
-        FILETOOL, "-d", "--mime-type", `./${p.base}`
+        FILETOOL, "-d", "--mime-type", `${p.dir}/${p.base}`
```
Comment (RealDolos) [inline lib/meta.js:311]: Do not change this either

**Review by RealDolos — CHANGES_REQUESTED** [review]
Things must work in node-14-lts at the very least.
node-12-lts and 16 a plus.

**Comment by voroneril** [comment]
All should be now fixed

**Review by RealDolos — APPROVED** [review]

**Pull request closed without merging**